streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -719,6 +723,8 @@ void runOnce() {
 
         final long pollLatency = pollPhase();
 
+        topologyMetadata.maybeWaitForNonEmptyTopology(() -> state);
```

Comment:
How about moving this ahead of `pollPhase()`? We are likely to be kicked out of the group while blocked waiting here, so it's better to be aware of that and re-join the group immediately, rather than doing the restore/etc still which may be all wasted work.

Reply:
Ack (although note that there's no wasted work on the restore phase since there's by definition nothing for the thread to do yet as it won't have been assigned any new tasks until it polls again). 

I don't think it really matters much where we put this for that reason, except for the case in which we start up with no topology -- then it's a waste to join the group in the first place, so we may as well wait until we receive something to work on. So yes, I'll move it back ahead of poll